lib/connections/service.go

```diff
@@ -336,176 +338,203 @@ func (s *service) handle(ctx context.Context) error {
+		// clamped by stdConnectionLoopSleep as we don't want to sleep too
+		// long (config changes might happen).
+		sleep := filterAndFindSleepDuration(nextDial, now)
 
-			if deviceCfg.Paused {
-				continue
-			}
+		// .. unless we are un the initial rampup time, in which case we
+		// override the sleep interval for the first few passes through the
+		// loop.
+		if isInitialRampup {
+			sleep = initialRampup
+			initialRampup *= 2
+		}
 
-			ct, connected := s.model.Connection(deviceID)
+		// ... while making sure not to loop to quickly either.
+		if sleep < minConnectionLoopSleep {
+			sleep = minConnectionLoopSleep
+		}
 
-			if connected && ct.Priority() == bestDialerPrio {
-				// Things are already as good as they can get.
-				continue
-			}
+		l.Debugln("Next connection loop in", sleep)
 
-			var addrs []string
-			for _, addr := range deviceCfg.Addresses {
-				if addr == "dynamic" {
-					if s.discoverer != nil {
-						if t, err := s.discoverer.Lookup(ctx, deviceID); err == nil {
-							addrs = append(addrs, t...)
-						}
-					}
-				} else {
-					addrs = append(addrs, addr)
-				}
-			}
+		select {
+		case <-time.After(sleep):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
 
-			addrs = util.UniqueTrimmedStrings(addrs)
+func (s *service) bestDialerPrio(cfg config.Configuration) int {
+	first := true
+	var bestDialerPrio int
+	for _, df := range dialers {
+		if df.Valid(cfg) != nil {
+			continue
+		}
+		if prio := df.Priority(); first || prio < bestDialerPrio {
+			first = false
+			bestDialerPrio = prio
+		}
+	}
+	return bestDialerPrio
+}
 
-			l.Debugln("Reconnect loop for", deviceID, addrs)
+func (s *service) dialDevices(ctx context.Context, now time.Time, cfg config.Configuration, bestDialerPrio int, nextDial map[string]time.Time, initial bool) {
+	for _, deviceCfg := range cfg.Devices {
+		// Don't attempt to connect to ourselves...
+		if deviceCfg.DeviceID == s.myID {
+			continue
+		}
 
-			dialTargets := make([]dialTarget, 0)
+		// Don't attempt to connect to paused devices...
+		if deviceCfg.Paused {
+			continue
+		}
 
-			for _, addr := range addrs {
-				// Use a special key that is more than just the address, as you might have two devices connected to the same relay
-				nextDialKey := deviceID.String() + "/" + addr
-				seen = append(seen, nextDialKey)
-				nextDialAt, ok := nextDial[nextDialKey]
-				if ok && initialRampup >= sleep && nextDialAt.After(now) {
-					l.Debugf("Not dialing %s via %v as sleep is %v, next dial is at %s and current time is %s", deviceID, addr, sleep, nextDialAt, now)
-					continue
-				}
-				// If we fail at any step before actually getting the dialer
-				// retry in a minute
-				nextDial[nextDialKey] = now.Add(time.Minute)
+		// Don't connect to devices that are already connected using the
+		// best possible dialer priority...
+		connection, connected := s.model.Connection(deviceCfg.DeviceID)
+		if connected && connection.Priority() == bestDialerPrio {
```

Comment:
Unrelated to this PR: This should be `>=`. If we e.g. listen on tcp but disabled dialing on tcp, we don't need to dial lower-prio methods when a tcp connection is already established.

Just saw that later on in the dialer loop we do use the `>=` condition between dialer prio and current connection prio.